### DISCUSSION TITLE
LMDB very slow with server/workbench

### DISCUSSION THREAD

**Discussion by maxstolze** [eclipse-rdf4j/rdf4j#3953] [Q&A] [answered]
We created a sail lmdb repository via console on the RDF4J server and tested it against native and memory store and it so 10 times slower.
This is our template:

```
#
# Rdf4j configuration template for a sail-lmdb repository
#
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix rep: <http://www.openrdf.org/config/repository#>.
@prefix sr: <http://www.openrdf.org/config/repository/sail#>.
@prefix sail: <http://www.openrdf.org/config/sail#>.
@prefix lmdb: <http://rdf4j.org/config/sail/lmdb#>.
@prefix sl: <http://www.openrdf.org/config/sail/lucene#>.
@prefix ssc: <http://rdf4j.org/config/sail/shacl#> .

[] a rep:Repository ;
   rep:repositoryID "{%Repository ID|memory%}" ;
   rdfs:label "{%Repository title|Max Test Repo%}" ;
   rep:repositoryImpl [
      rep:repositoryType "openrdf:SailRepository" ;
      sr:sailImpl [
         sail:sailType "rdf4j:ShaclSail" ;
         ssc:parallelValidation true;
         sail:delegate [
          sail:sailType "rdf4j:LmdbStore"
          ];
      ];
   ].
```

What are we doing wrong/missing?
(Already fixed the dependency LWJGL bug before)

Thanks in advance!

**Comment by hmottestad** [comment]
Is it 10x slower than the NativeStore?

What exactly is 10x slower? Loading data? Querying? Deleting? Updating?

If you use the LmdbStore directly, without the ShaclSail, is it still slow?

**Reply by maxstolze** [reply]
Yes (at least Sail + MemoryStore)
Querying/Loading the Data.
Still the same without ShaclSail.

**Reply by kenwenzel** [reply]
@maxstolze 
- Have you also tried it with a standalone program? 
- Can you provide an example dataset?
- What type of storage are you using HDD or SSD? Is it a network-attached storage?

Maybe the way how connections are managed within the server has an influence on the performance...

**Reply by hmottestad** [reply]
The LMDB store is persistent (ACID) so it is expected to be slower than the MemoryStore. The MemoryStore can enabled with persistence, but it’s only a best-effort and not ACID. 

The NativeStore is typically slower than the LMDB store, at least in our benchmarks. Query statistics is implemented in a different way though, so some queries might result in different  query plans and as such better or worse because of that.

**Reply by hmottestad** [reply]
@kenwenzel is the developer of the LMDB store. So he knows most about how it works.

**Reply by kenwenzel** [reply]
@hmottestad  Do we have some setup for benchmarking with RDF4J server?

**Reply by hmottestad** [reply]
I always use docker for testing.

**Reply by hmottestad** [reply]
This script: https://github.com/eclipse/rdf4j/blob/main/docker/build.sh

**Reply by kenwenzel** [reply]
Thank you. I think we'll have to run our standard benchmarks also against the server.

**Reply by hmottestad** [reply]
We also have a maven+jetty+junit test server. It's not particularly streamlined unfortunately since you need to build the war file from maven every time you change something. I'm also not sure if it's possible to easily debug or profile. 

Here is an example of using it to test SHACL validation:  

https://github.com/eclipse/rdf4j/blob/main/tools/server/src/test/java/org/eclipse/rdf4j/http/server/ShaclValidationReportIT.java

**Reply by maxstolze** [reply]
Thank you for the input!

So first of all the template should be correct to create the LmdbStore via console on the RDF4JServer and we are not missing something?

@kenwenzel
 - Only with the RDF4JServer and against the Workbench included templates
 - Need to check this
 - SSD and all (Memory/Native/LMDB) on the same RDF4JServer running on Tomcat 9 on local machine

@hmottestad 
Checked again and Native + MemoryStore are significant faster.

Here the System Information if this helps:

![image](https://user-images.githubusercontent.com/32457902/171794164-e815bfa6-5f8b-44e1-96f8-8b28f180eed4.png)

**Reply by kenwenzel** [reply]
@maxstolze Thank you for the information. I will try to setup some benchmarks next week.
Maybe you can also give us some more details about your dataset and the queries?

**Reply by hmottestad** [reply]
Maybe 256MB of ram is too little?

**Reply by kenwenzel** [reply]
LMDB uses off-heap memory and relies on the OS functions for paging. If the system RAM size is too small then it may also be the source for the bad performance.

**Reply by maxstolze** [reply]
Okay, so giving it more memory (2GB) doubles the speed (36 seconds before | 16 seconds now)

**Reply by kenwenzel** [reply]
But MemoryStore and NativeStore only take 1.6 seconds?

**Reply by maxstolze** [reply]
They took around 3 seconds before (when it was over 30 seconds) even with only 256MB of ram.

**Reply by kenwenzel** [reply]
Can you describe a bit what you are doing? How do you insert/query the data and how much data are you inserting?
Is the insertion speed slow or the query speed?

You could also try to set higher value and triple DB sizes:
https://rdf4j.org/javadoc/4.0.1/org/eclipse/rdf4j/sail/lmdb/config/LmdbStoreConfig.html

It could be the case that under Windows the autogrow functionality is slow. I've only tested it under Linux.

**Reply by maxstolze** [reply]
Also thought the `autogrow` could be a factor, but the data insertion is not slow.
It's when querying the data. Maybe our queries only work good with GraphDB, MemoryStore & NativeStore? But can't imagine this. 

So what we are doing is running everything locally for developement and a typical setup:
- RDF4J Server in standalone Tomcat
- Spring + RDF4J Backend communicating with the RDF4J Server + SAIL LMDB Repository
- Frontend communicating with the backend

Tested so far:

- GraphDB
- MemoryStore
- NativeStore
- LMDB

And only with the lmdb setup the queries took ages, but thats actually the setup we wanted to use in the first place at it theoretically semed to be the one with the best performance.

**Reply by kenwenzel** [reply]
We had a severe bug regarding the query optimization in the LMDB store:
https://github.com/eclipse/rdf4j/pull/3869

Maybe your current version does not yet include this fix?
Are you using the 4.0.1 release version?

**Reply by kenwenzel** [reply]
Can you please provide example data and one of the slow queries?

**Reply by maxstolze** [reply]
Yes using the 4.0.1

**Reply by hmottestad** [reply]
The query planner relies on good statistics to create an optimal query. All three stores calculate their statistics differently.

If you are able to run things locally (without the server) you can compare the query plans. https://rdf4j.org/documentation/programming/repository/#explaining-queries

**Reply by fkleedorfer** [reply]
The problem is that we are trying to improve the performance of a specific case in our Web application (that has an RDF4J backend). That case is a user-initiated copy process involving many queries and updates. We will have a hard time comparing query plans right now. However, we'll do some drilling to find out if there is a particularly slow query, and if we find one/some, post them here along with some dummy testdata. We can't provide the actual test data, I am afraid.

**Reply by kenwenzel** [reply]
OK. Maybe there are also other issues related to opening the LMDB store and/or transactions. I'll wait for your input.

**Reply by maxstolze** [reply]
Okay. Now we tracked it down to the actual operation and could compare LMDB Store with Native Store.

LMDB:
![lmdbStats](https://user-images.githubusercontent.com/32457902/172613387-82e0f0f4-5a2a-475c-84e8-0293c1c16ef4.PNG)

Native:
![nativeStatus](https://user-images.githubusercontent.com/32457902/172613407-0292d83a-b97e-4a5e-a8a6-8e3a088c55e3.PNG)

The operation directly logged with JConsole/JMX:
```
PREFIX mms: <https://mms.researchstudio.at/mms#> 
PREFIX s: <http://schema.org/> 
SELECT DISTINCT ?id ?f_copyOf ?f_standardId ?f_name ?dateCreated ?dateModified ?f_private ?f_description ?creator ?editor ?f_featureKind ?class ?quantityKind ?quantityKindLabel ?unit ?unitLabel ?allowMultiple ?optionId ?optionValue ?optionDescription ( CONCAT( IF( BOUND( ?organizationId ), CONCAT( "/organizations/", REPLACE( STR( ?organizationId ), "urn:uuid:|/|#|:|\\\\.", "" ) ), CONCAT( "/projects/", REPLACE( STR( ?projectId ), "urn:uuid:|/|#|:|\\\\.", "" ) ) ), "/standards/", REPLACE( STR( ?f_standardId ), "urn:uuid:|/|#|:|\\\\.", "" ), "/features/", REPLACE( STR( ?id ), "urn:uuid:|/|#|:|\\\\.", "" ) ) AS ?path ) 
WHERE { 
?id a mms:Feature ;
     s:name ?f_name ;
     mms:private ?f_private ;
     s:dateCreated ?dateCreated ;
     s:dateModified ?dateModified ;
     s:creator ?creator ;
     s:editor ?editor . 
?id mms:belongsToStandard ?f_standardId . 
OPTIONAL { ?f_standardId mms:belongsToOrganization ?organizationId . } 
OPTIONAL { ?f_standardId mms:belongsToProject ?projectId . } OPTIONAL { ?id s:description ?f_description . } 
OPTIONAL { ?id mms:copyOf ?f_copyOf . } ?id mms:dataType ?f_featureKind . { ?f_featureKind a ?class . 
OPTIONAL { ?f_featureKind a ?class . 
    FILTER ( ( ?f_featureKind != <http://www.linkedmodel.org/schema/dtype#SimpleCodeList> && ?f_featureKind != mms:DataType ) ) 
} 
OPTIONAL { ?f_featureKind mms:quantityKind ?quantityKind ;     mms:unit ?unit . ?quantityKind <http://www.w3.org/2000/01/rdf-schema#label> ?quantityKindLabel . ?unit <http://www.w3.org/2000/01/rdf-schema#label> ?unitLabel . } 
OPTIONAL { ?f_featureKind mms:allowMultiple ?allowMultiple . } 
OPTIONAL { ?f_featureKind <http://www.linkedmodel.org/schema/dtype#hasMember> ?optionId . ?optionId <http://www.linkedmodel.org/schema/dtype#value> ?optionValue . 
    OPTIONAL { ?optionId s:description ?optionDescription . } 
}
FILTER ( ( ?f_featureKind IN ( mms:StringType, mms:BooleanType, mms:ReferenceType ) || ?class IN ( mms:NumericType, mms:EnumerationType ) ) ) } ?root_feature mms:belongsToStandard ?root_standard ;     mms:copyOf ?id . 
FILTER ( !( ( BOUND( ?organizationId ) && BOUND( ?projectId ) ) ) ) } 

```


We can not provide the actual data we used for this test, but we have some test data on which the operation can be tested.

[simple-testdata.ttl](https://gist.github.com/maxstolze/87ce3b98b2c7d883c420f54d93ec38cf)
[mms.ttl](https://gist.github.com/maxstolze/ebd46a7a3c9b716047ba6e9b90c01352)

**Reply by kenwenzel** [reply]
Thank you. So the problem seems to be related to query optimization. I'll take a look!

**Reply by hmottestad** [reply]
> Thank you. So the problem seems to be related to query optimization. I'll take a look!

I made a few different evaluation statistics implementations for the extensible store. 

https://github.com/eclipse/rdf4j/tree/main/core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics

There is a dynamic one that aims at tracking added and removed statements with HLL (HyperLogLog).  That one may be particularly useful. 

https://github.com/eclipse/rdf4j/blob/main/core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java

**Reply by kenwenzel** [reply]
@hmottestad Thank you for pointing this out. I already took a look at your statistics implementation. Do you see some way to persist the HLL instances?
I also thought about implementing some kind of counting bloom filter which can be persisted within an LMDB db.

**Reply by hmottestad** [reply]
There is a toBytes(...) method: https://github.com/aggregateknowledge/java-hll/blob/1f4126e79a85b79581460c75bf1c1634b8a7402d/src/main/java/net/agkn/hll/HLL.java#L880

**Reply by hmottestad** [reply]
Fyi (from a quick google search):

> HyperLogLog is similar to a Bloom filter internally as it runs items through a non-cryptographic hash and sets bits in a form of a bitfield. Unlike a Bloom filter, HyperLogLog keeps a counter of items that is incremented when new items are added that have not been previously added. This provides a very low error rate when estimating the unique items (cardinality) of a set.

**Reply by kenwenzel** [reply]
> 
> We can not provide the actual data we used for this test, but we have some test data on which the operation can be tested.
> 
> [simple-testdata.ttl](https://gist.github.com/maxstolze/87ce3b98b2c7d883c420f54d93ec38cf) [mms.ttl](https://gist.github.com/maxstolze/ebd46a7a3c9b716047ba6e9b90c01352)

@maxstolze @fkleedorfer Do you have any rules to scale/duplicate the test data to get more triples?

**Reply by fkleedorfer** [reply]
I can create a bigger dataset, but I am not certain the behaviour will be the same. I'll get back to you when I have one.

**Reply by kenwenzel** [reply]
Maybe you can also provide some kind of generator function for Java? I thought about using your example data and simpling duplicating the items (while following certain rules).

**Reply by fkleedorfer** [reply]
here it is. I can make it bigger if needed.
https://gist.github.com/fkleedorfer/ffdb53f2419ed750158680b250ba84b9

**Reply by fkleedorfer** [reply]
> Maybe you can also provide some kind of generator function for Java? I thought about using your example data and simpling duplicating the items (while following certain rules).

I don't see an easy way for us to do that. It's much faster to generate the data on our side as needed.

**Reply by kenwenzel** [reply]
Thank you for the test data. I've implemented a test for both the NativeStore and the LmdbStore:
https://github.com/kenwenzel/rdf4j/tree/GH-3982-lmdb-query-optimization

Unfortunately, I get almost the same results for your example query. Do you bind some of the variables when you execute the query?

LmdbStore:
![lmdbstore](https://user-images.githubusercontent.com/177272/173290465-28676734-2841-48a0-8e22-4e4b8d074001.svg)

NativeStore:
![nativestore](https://user-images.githubusercontent.com/177272/173290470-422281a2-d5c2-4d8c-844f-af1dd158cc38.svg)

**Reply by fkleedorfer** [reply]
> Do you bind some of the variables when you execute the query?

@maxstolze correct me if I am wrong - but it seems to me that this query can only be executed in our code with a binding for `?root_standard` (and nothing else).

The dataset I provided will not produce interesting results there because the 'standard' in question has not been copied. Here is another dataset: https://gist.github.com/fkleedorfer/3b03ab4612f470b4cf99b3bf26bd575b

... **Meant to be added to the first one.**

It contains a 'standard' that has been copied. (so there are `?x mms:copyOf ?y` triples in there. I think binding `?root_standard` to `urn:uuid:fed35f18-bd2e-4cfe-a6ea-bbb351b49e75` might lead to our observed behaviour.

**Reply by kenwenzel** [reply]
OK, this is really helpful. I'll first try to improve the existing algorithm for computing the cardinalities before moving to something like HLL.

**Comment by kenwenzel** [comment]
@maxstolze Which indexes are you using with the LMDB store. The default configuration?

**Reply by fkleedorfer** [reply]
We did not change the indexing.

**Reply by kenwenzel** [reply]
OK. Maybe you can try with "spoc,ospc,psoc" until the issue ist fixed.

**Comment by kenwenzel** [comment, marked as answer]
@fkleedorfer @maxstolze Maybe you can try https://github.com/eclipse/rdf4j/pull/4183 with your data?

**Reply by kenwenzel** [reply]
The fix is included in RDF4J 4.2.0.

**Reply by maxstolze** [reply]
Thank you! It's now as fast as NativeStore with our setup!

**Reply by kenwenzel** [reply]
@maxstolze Thank you for the information. How many triples has your data set?

**Reply by maxstolze** [reply]
![image](https://user-images.githubusercontent.com/32457902/193223767-a3877220-26cc-4bc5-b689-0ce7de57d52c.png)

**Reply by kenwenzel** [reply]
Ah ok, that is not much. Probably the LMDB store mainly pays off if the statements don't fit completely into RAM.
Nonetheless, it would be interesting if you could collect some profiling information with VisualVM.
